Fix equivalent area calculation

snow54 opened the pull request:
## Proposed Changes
*Give a brief overview of your contribution here in a few sentences.*
Since #724, equivalent area calculation and its use as objective function when EQUIV_AREA is set to YES have not been functioning. I have moved `SpecialOutput_SonicBoom` from output_structure_legacy.cpp to CFlowOutput.cpp and called it out in CFlowCompOutput.cpp similarly to inverse design Cp function mentioned in #1311.  This PR fixes #841.

This is my first PR on github, so let me know if I should fix anything.

Mesh file and TargetEA.dat
[NACA64206_EA.zip](https://github.com/su2code/SU2/files/6839247/NACA64206_EA.zip)

## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*
#724 
#1311 
#841 

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [x] My contribution is commented and consistent with SU2 style.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

pcarruscag commented on Common/include/option_structure.hpp:
```diff
@@ -1489,6 +1489,7 @@ static const MapType<std::string, ENUM_OBJECTIVE> Objective_Map = {
   MakePair("MOMENT_Y", MOMENT_Y_COEFFICIENT)
   MakePair("MOMENT_Z", MOMENT_Z_COEFFICIENT)
   MakePair("EQUIVALENT_AREA", EQUIVALENT_AREA)
+  //MakePair("EQUIV_AREA", EQUIVALENT_AREA)
```
Avoid leaving commented-out code, delete what you don't need.

snow54 replied:
Deleted commented out codes.

pcarruscag commented on SU2_CFD/include/output/CFlowOutput.hpp:
```diff
@@ -88,6 +88,19 @@ class CFlowOutput : public CFVMOutput{
+   * \param[in] geometry - Geometrical definition of the problem.
+   * \param[in] config - Definition of the particular problem.
+   */
+  void Set_NearfieldInverseDesign(CSolver *solver, const CGeometry *geometry, const CConfig *config);
```
:+1: On the const correctness

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,647 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+}
+
+void CFlowOutput::Set_NearfieldInverseDesign(CSolver *solver, const CGeometry *geometry, const CConfig *config){
+  // Transferred from output_structure_legacy.cpp
```
```suggestion
```

snow54 replied:
Deleted.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,647 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  /*--- Create an array with all the coordinates, points, pressures, face area,
+   equivalent area, and nearfield weight ---*/
+
+  Xcoord = new su2double[nVertex_NearField];
+  Ycoord = new su2double[nVertex_NearField];
+  Zcoord = new su2double[nVertex_NearField];
+  AzimuthalAngle = new short[nVertex_NearField];
+  IdPoint = new unsigned long[nVertex_NearField];
+  IdDomain = new unsigned long[nVertex_NearField];
+  Pressure = new su2double[nVertex_NearField];
+  FaceArea = new su2double[nVertex_NearField];
```
If you want to modernize this a little you can use std::vector.
Have a look at CFlowOutput::SetAnalyzeSurface if you are interested.

snow54 replied:
I have converted array to std::vector.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,647 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+#else
+
+  int nProcessor;
+  SU2_MPI::Comm_size(SU2_MPI::GetComm(), &nProcessor);
```
`size` is a member of the output I think

snow54 replied:
Can you elaborate a bit more on this?

pcarruscag replied:
Most of SU2's large classes store the mpi rank and mpi size as a member variable and initialize them in the constructor via GetRank() and GetSize().
COuput already has a member called "size" that is equivalent to this "nProcessor".
This was maybe needed in the legacy output, but not in the new one, that's all.

snow54 replied:
Thanks. I have replaced it with `size`.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,647 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  su2double *Buffer_Send_Pressure = new su2double [MaxLocalVertex_NearField];
+  su2double *Buffer_Send_FaceArea = new su2double[MaxLocalVertex_NearField];
+
+  su2double *Buffer_Receive_Xcoord = NULL;
+  su2double *Buffer_Receive_Ycoord = NULL;
+  su2double *Buffer_Receive_Zcoord = NULL;
```
`nullptr` instead of NULL (but std::vector would save all that)

snow54 replied:
See below.
https://github.com/su2code/SU2/pull/1329#discussion_r674043743

TobiKattmann replied:
Referencing a comment 😮 ... i didn't know that was possible

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,647 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+     this first implementation requires a complete table (same as the original
+     EA table). so... no interpolation. ---*/
+
+    vector<vector<su2double> > TargetArea_PhiAngle_Trans;
```
su2activematrix for things that look like matrices

snow54 replied:
`su2activematrix` doesn't appear to have a method equivalent to `push_back`. Do I have to resize the matrix and insert values one by one?

pcarruscag replied:
Oh if it is using push_back you can leave it like it is.

snow54 replied:
OK, I'll leave it as is then.

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -845,7 +845,8 @@ void CEulerSolver::Set_MPI_Nearfield(CGeometry *geometry, CConfig *config) {
 
       SU2_MPI::Isend(&nPointTotal_s[iDomain], 1, MPI_UNSIGNED_LONG, iDomain, iDomain, SU2_MPI::GetComm(), &req);
-      SU2_MPI::Request_free(&req);
+      SU2_MPI::Wait(&req,&status);
+      //SU2_MPI::Request_free(&req);
```
Have you stress tested that this does not create deadlocks?

snow54 replied:
No, but `SU2_MPI::Request_free` creates errors shown below during adjoint calculation, so `SU2_MPI::Wait` was a workaround. Is there a better way?

Error in function AMPI_Request_free (../externals/medi/include/medi/ampi/async.hpp:249)
The message is: Freeing a handle that is not finish with wait, waitall, etc..

pcarruscag replied:
I see... That routine was written to use non-blocking comms (whether it needs that or not is a different story).
By waiting right after the Send/Recv you make it equivalent to a blocking communication.
If the non-blocking was indeed necessary this may cause the code to dead-lock because a Send will not be matched with the right Recv.
It would be good to test this with gross-oversubscribtion, e.g. 17 ranks on a system with 4 cores while playing a youtube video.

snow54 replied:
I already encountered an issue without doing the test you mentioned. I changed a number of cores to use and the adjoint calculation seems to stop there. As I'm not very familiar with MPI, I was doing some research on MPI but I have not been able to solve this so far.

pcarruscag replied:
Ok. Based on CGeometry::PostP2PRecvs (and similar).
You need to create a vector of all the requests (one per message).
You launch all the IRecv first.
Then you launch all the sends.
Finally you "Waitall" on the receive requests.

snow54 replied:
I tried to write something similar to `CGeometry::PostP2PRecvs`, `CGeometry::PostP2PSends`, and `CGeometry::CompleteComms` but I cannot get it to work for some reason. Segmentation fault occurs during Waitall.  Do you have any idea on what is possible causing this?

```
*** Process received signal ***
Signal: Segmentation fault (11)
Signal code: Address not mapped (1)
Failing at address: 0x1520
*** An error occurred in MPI_Waitall
*** reported by process [2490630145,3]
*** on communicator MPI_COMM_WORLD
*** MPI_ERR_REQUEST: invalid request
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
```

pcarruscag replied:
I will try to help when I have some time.

snow54 replied:
Thanks. I'm still working on it (slowly).

snow54 replied:
After learning how MPI works, MPI appears to be working now but the program stops with the following messages. Google says it is related to memory issues where you delete the same pointer twice or you try to access address outside of a vector, but I cannot find such issues in my codes. Can you help me identify an issue?

```
*** Process received signal ***
Signal: Aborted (6)
Associated errno: Unknown error 32766 (32766)
Signal code:  (-543634320)
```
and
```
free(): invalid next size (fast)
```

pcarruscag replied:
You can try using valgrind. --buildtype=debugoptimized, then mpirun -n "n" valgrind SU2_CFD ... and you should get messages if you are double-freeing or accessing out of bounds. But if it is happening inside MPI it will be difficult.

That entire function was really badly written, it should be doing an AllToAll for the counts followed by an AllToAllV for the data, not all this complication with Sends and Recvs.

snow54 replied:
Thanks! valgrind found an issue and I fixed it. It seems to be working now. I have introduced some variables for debug purpose, so I will do a little clean-up later.

snow54 replied:
I think I'm now done with this comments.

pcarruscag commented on SU2_PY/SU2/io/historyMap.py:
```diff
@@ -451,7 +451,7 @@
                 'GROUP': 'AERO_COEFF',
                 'HEADER': 'CEff',
                 'TYPE': 'COEFFICIENT'},
- 'EQUIV_AREA': {'DESCRIPTION': 'Equivalent area',
+ 'EQUIVALENT_AREA': {'DESCRIPTION': 'Equivalent area',
```
Did you run the script (updateHistoryMap or something like that) to update this file or did it by hand?

snow54 replied:
I updated it by hand. Now I know how to use updateHistoryMap.py, I used this script to update historyMap.py. It changed things that I didn't touch. Is it OK to update them?

Change
- BGS_TEMPERATURE

Added
- STREAMWISE_DP
- STREAMWISE_HEAT
- STREAMWISE_MASSFLOW

pcarruscag replied:
It should be... @TobiKattmann I think these are "your" outputs, any objections?

TobiKattmann replied:
No objections, I forgot to do that in #773 obviously 😞 ... for bgs-temperature I changed the explaining string in the code which is automatically ported.

snow54 replied:
OK, thanks.

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,434 @@
+RESTART_SOL= NO 
+%
+% Write binary restart files (YES, NO)
+%WRT_BINARY_RESTART= NO
+%Commented out for v7
+%
```
No comments here either, delete everything superfluous, make the sole focus of this config the equivalent area.
Only make detailed comments on the feature you are testing.

snow54 replied:
I have separated the config file into direct and adjoint calculations and deleted everything that is not necessary.

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,434 @@
+
+% ---------------------- REFERENCE VALUE DEFINITION ---------------------------%
+%
+% Reference origin for moment computation
+REF_ORIGIN_MOMENT_X = 0.25
+REF_ORIGIN_MOMENT_Y = 0.00
+REF_ORIGIN_MOMENT_Z = 0.00
+%
+% Reference length for pitching, rolling, and yaMAIN_BOX non-dimensional moment
+REF_LENGTH= 1
+%
+% Reference area for force coefficients (0 implies automatic calculation)
+REF_AREA= 0
+%
+% Flow non-dimensionalization (DIMENSIONAL, FREESTREAM_PRESS_EQ_ONE,
+%                              FREESTREAM_VEL_EQ_MACH, FREESTREAM_VEL_EQ_ONE)
+REF_DIMENSIONALIZATION= FREESTREAM_PRESS_EQ_ONE
```
```suggestion
REF_ORIGIN_MOMENT_X = 0.25
REF_ORIGIN_MOMENT_Y = 0.00
REF_ORIGIN_MOMENT_Z = 0.00
REF_LENGTH= 1
REF_AREA= 0
REF_DIMENSIONALIZATION= FREESTREAM_PRESS_EQ_ONE
```
For example

snow54 replied:
See below.
https://github.com/su2code/SU2/pull/1329#discussion_r673786199

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,434 @@
+REF_DIMENSIONALIZATION= FREESTREAM_PRESS_EQ_ONE
+
+% ------------------------ SUPERSONIC SIMULATION ------------------------------%
+%
+% Evaluate equivalent area on the Near-Field (NO, YES)
+EQUIV_AREA= YES
```
Add more detail please, including how you should set up the markers for this function.

snow54 replied:
I have added brief instruction.

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,434 @@
+%
+% Adaptive CFL number (NO, YES)
+CFL_ADAPT= NO
+%
+% Parameters of the adaptive CFL number (factor down, factor up, CFL min value,
+%                                        CFL max value )
+CFL_ADAPT_PARAM= ( 0.1, 2.0, 3, 100.0, 0.001 )
+%
+% Runge-Kutta alpha coefficients
+RK_ALPHA_COEFF= ( 0.66667, 0.66667, 1.000000 )
+%
+% Number of total iterations
```
```suggestion
```

snow54 replied:
See below.
https://github.com/su2code/SU2/pull/1329#discussion_r673786199

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,434 @@
+
+% -------------------------- MULTIGRID PARAMETERS -----------------------------%
+%
+% Multi-Grid Levels (0 = no multi-grid)
+MGLEVEL= 0
+%
+% Multi-grid cycle (V_CYCLE, W_CYCLE, FULLMG_CYCLE)
+MGCYCLE= W_CYCLE
+%
+% Multi-Grid PreSmoothing Level
+MG_PRE_SMOOTH= ( 1, 2, 3, 3 )
+%
+% Multi-Grid PostSmoothing Level
+MG_POST_SMOOTH= ( 0, 0, 0, 0 )
+%
+% Jacobi implicit smoothing of the correction
+MG_CORRECTION_SMOOTH= ( 0, 0, 0, 0 )
+%
+% Damping factor for the residual restriction
+MG_DAMP_RESTRICTION= 0.9
+%
+% Damping factor for the correction prolongation
+MG_DAMP_PROLONGATION= 0.9
+
+
```
```suggestion
MGLEVEL= 0
```
And so on

snow54 replied:
See below.
https://github.com/su2code/SU2/pull/1329#discussion_r673786199

pcarruscag reviewed (commented):
Welcome to SU2 and thank you for the contribution!
Did you decide to open the PR based on the conference? (asking for statistical purposes :smile: )
Some comments below to improve this old code a little (if you want).
And other more important aspects of adding test cases.

If you want to join the dev meeting this Weds at 3pm UK (https://meet.jit.si/SU2_DevMeeting) to let us know of other things you are working on, or just to meet the current "voices" behind SU2 you are most welcome.

Cheers!

pcarruscag commented on SU2_CFD/include/solvers/CFVMFlowSolverBase.inl:
```diff
@@ -2878,6 +2878,8 @@ su2double CFVMFlowSolverBase<V,R>::EvaluateCommonObjFunc(const CConfig& config)
     case INVERSE_DESIGN_HEATFLUX:
       objFun += weight * Total_HeatFluxDiff;
       break;
+    case EQUIVALENT_AREA:
+      objFun += weight*Total_CEquivArea;
```
```suggestion
    case EQUIVALENT_AREA:
      objFun += weight*Total_CEquivArea;
      break;
```

snow54 replied:
Thanks.  I missed this.

snow54 commented:
@pcarruscag 
Thank you for your quick review. I'll address your comments tomorrow or later.

I decided to open the PR because I wanted to use the function introduced by #1020. I kept using v6 but was having hard time getting adjoint to converge. Nevertheless, I looked at your presentation for the conference (I could not listen to your talk due to time zone) and it encouraged me too.

It would be interesting to join the dev meeting but it's a bit hard again due to time zone. Thanks for asking though.

pcarruscag commented:
Great!
I see, I will propose different times for the meetings on the Slack channel (https://join.slack.com/t/su2devteam/shared_invite/zt-af0uuqf8-8XNExKMV9G~UVsnkvi5uVA), even for some folks in the US it is very early.
Join the channel if you would like and you can help to pick better times, we want all aspects of SU2 to be accessible worldwide.

snow54 commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,647 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  /*--- Compute the total number of points on the near-field ---*/
+
+  nVertex_NearField = 0;
+  for (iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++)
```
I have deleted #ifndef HAVE_MPI sections. When I compile the codes without MPI, adjoint calculation stops with the message below. I have tested with the develop branch too and this error occurs as well. It appears that this error is unrelated to the changes I made.
```
Error in "void CSymmetricMatrix::CholeskyDecompose()": 
-------------------------------------------------------------------------
LLT factorization failed.
```

pcarruscag replied:
That is coming from the QuasiNewton feature. We can talk more in today's meeting although I have a bit of overlap with another one... I'm glad you can join it btw.

snow54 replied:
When I compiled with OpenBLAS and LAPACK as discussed in the dev meeting, the error above didn't occur even if I compile without MPI. Thanks for your suggestion.

snow54 commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,434 @@
+%                                                                              %
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+% ------------- DIRECT, ADJOINT, AND LINEARIZED PROBLEM DEFINITION ------------%
+%
+% Physical governing equations (EULER, NAVIER_STOKES)
+SOLVER= EULER
```
Opened a PR [65](https://github.com/su2code/TestCases/pull/65) on TestCases repository. The only existing mesh with MARKER_NEARFIELD filled out or EQUIV_AREA set to YES is mesh_BIPARABOLIC_sup.su2 but it is in 2D. I don't think equivalent area calculation is very useful in 2D, so I want to make sure that my modification works on 3D mesh. I hope a mesh with 126k nodes is not too big.

I'll make modifications to parallel_regression.py later.

snow54 replied:
I have added test cases in parallel_regression.py and parallel_regression_AD.py.

WallyMaier commented on SU2_CFD/include/solvers/CEulerSolver.hpp:
```diff
@@ -840,7 +840,7 @@ class CEulerSolver : public CFVMFlowSolverBase<CEulerVariable, ENUM_REGIME::COMP
    * \return Value of the Equivalent Area coefficient (inviscid + viscous contribution).
    */
-  inline su2double GetTotal_CEquivArea() const final { return Total_CEquivArea; }
+  //inline su2double GetTotal_CEquivArea() const final { return Total_CEquivArea; }
```
Are these unused?  If so, can they be removed?

snow54 replied:
I have deleted them.

WallyMaier commented on SU2_CFD/include/solvers/CFVMFlowSolverBase.hpp:
```diff
@@ -2118,6 +2126,8 @@ class CFVMFlowSolverBase : public CSolver {
    */
   inline su2double GetTotal_HeatFluxDiff() const final { return Total_HeatFluxDiff; }
 
+  inline su2double GetTotal_CEquivArea() const final { return Total_CEquivArea; }
```
Can you add a brief for this?

snow54 replied:
I have moved the description from original location.

WallyMaier commented on SU2_CFD/src/output/CFlowCompOutput.cpp:
```diff
@@ -226,7 +226,7 @@ void CFlowCompOutput::SetHistoryOutputFields(CConfig *config){
   /// BEGIN_GROUP: EQUIVALENT_AREA, DESCRIPTION: Equivalent area.
   /// DESCRIPTION: Equivalent area
-  AddHistoryOutput("EQUIV_AREA",   "CEquiv_Area",  ScreenOutputFormat::SCIENTIFIC, "EQUIVALENT_AREA", "Equivalent area", HistoryFieldType::COEFFICIENT);
+  /// AddHistoryOutput("EQUIV_AREA",   "CEquiv_Area",  ScreenOutputFormat::SCIENTIFIC, "EQUIVALENT_AREA", "Equivalent area", HistoryFieldType::COEFFICIENT);
```
Can we removed this?

snow54 replied:
I have deleted this.

pcarruscag commented on SU2_CFD/include/solvers/CFVMFlowSolverBase.hpp:
```diff
@@ -2091,6 +2092,13 @@ class CFVMFlowSolverBase : public CSolver {
+   * \brief Set the value of the Equivalent Area coefficient.
+   * \param[in] val_equiv - Value of the Equivalent Area coefficient.
+   */
+
```
```suggestion
```

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_adjoint.cfg:
```diff
@@ -0,0 +1,159 @@
+REF_DIMENSIONALIZATION= FREESTREAM_PRESS_EQ_ONE
+
+% ------------------------ SUPERSONIC SIMULATION ------------------------------%
+% MARKER_NEARFIELD needs to be defined on a circumferential boundary within
+% calculation domain so that it captures pressure disturbance from the model.
+% The boundary should have a structured grid with the same number of nodes
+% along each azimuthal angle.
```
Perfect! Thanks!

TobiKattmann replied:
@snow54 please add/update the explanation in config_template.cfg in the code root. in case the option is not present please add the option

snow54 replied:
The same explanation as one in configuration file for the test case has been added.

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_adjoint.cfg:
```diff
@@ -0,0 +1,159 @@
+%
+NUM_METHOD_GRAD= GREEN_GAUSS
+
+% Objective function in gradient evaluation  (DRAG, LIFT, SIDEFORCE, MOMENT_X,
+%                                             MOMENT_Y, MOMENT_Z, EFFICIENCY,
+%                                             EQUIVALENT_AREA, NEARFIELD_PRESSURE,
+%                                             FORCE_X, FORCE_Y, FORCE_Z, THRUST,
+%                                             TORQUE, FREE_SURFACE, TOTAL_HEATFLUX,
+%                                             MAXIMUM_HEATFLUX, INVERSE_DESIGN_PRESSURE,
+%                                             INVERSE_DESIGN_HEATFLUX)
```
```suggestion
% Objective function in gradient evaluation
```
We are trying to avoid listing options in the examples because it makes it difficult to keep everything up to date.
We use the config_template to list the options, and the examples to highlight / test particular ones.
At the moment if you run SU2 and ask for a "wrong" option it will print all the valid possibilities.

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_direct.cfg:
```diff
@@ -0,0 +1,160 @@
+%
+% ----------------- ADJOINT-FLOW NUMERICAL METHOD DEFINITION ------------------%
+%
+CONV_NUM_METHOD_ADJFLOW= HLLC
+MUSCL_ADJFLOW=YES
+RELAXATION_FACTOR_ADJOINT= 1.0
+CFL_REDUCTION_ADJFLOW= 0.5
+TIME_DISCRE_ADJFLOW= EULER_IMPLICIT
+QUASI_NEWTON_NUM_SAMPLES= 20
+MG_ADJFLOW = NO
```
```suggestion
RELAXATION_FACTOR_ADJOINT= 1.0
QUASI_NEWTON_NUM_SAMPLES= 20
```
The options I removed are only used by the continuous adjoint. The discrete adjoint uses the normal numerical method definitions.

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_direct.cfg:
```diff
@@ -0,0 +1,160 @@
+%
+% Physical governing equations (EULER, NAVIER_STOKES)
+SOLVER= EULER
+%
+% Mathematical problem (DIRECT, CONTINUOUS_ADJOINT, DISCRETE_ADJOINT)
+MATH_PROBLEM= DIRECT
```
```suggestion
```
If this is the only difference between the two config files you can remove it.
If MATH_PROBLEM is omited, SU2 deduces the type of problem from the executable, SU2_CFD -> DIRECT, SU2_CFD_AD -> DISCRETE_ADJOINT

snow54 replied:
I have consolidated two configuration files into one.

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_direct.cfg:
```diff
@@ -0,0 +1,160 @@
+SURFACE_ADJ_FILENAME= surface_adjoint
+HISTORY_WRT_FREQ_INNER= 30
+SCREEN_WRT_FREQ_INNER= 1
+SCREEN_OUTPUT= (INNER_ITER, RMS_DENSITY, EQUIVALENT_AREA)
```
:+1:

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_direct.cfg:
```diff
@@ -0,0 +1,160 @@
+% ----------------------- SLOPE LIMITER DEFINITION ----------------------------%
+%
+SLOPE_LIMITER_FLOW=VAN_ALBADA_EDGE
+SLOPE_LIMITER_ADJFLOW=VAN_ALBADA_EDGE
```
```suggestion
```

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_direct.cfg:
```diff
@@ -0,0 +1,160 @@
+%
+NUM_METHOD_GRAD= GREEN_GAUSS
+
+% Objective function in gradient evaluation  (DRAG, LIFT, SIDEFORCE, MOMENT_X,
+%                                             MOMENT_Y, MOMENT_Z, EFFICIENCY,
+%                                             EQUIVALENT_AREA, NEARFIELD_PRESSURE,
+%                                             FORCE_X, FORCE_Y, FORCE_Z, THRUST,
+%                                             TORQUE, FREE_SURFACE, TOTAL_HEATFLUX,
+%                                             MAXIMUM_HEATFLUX, INVERSE_DESIGN_PRESSURE,
+%                                             INVERSE_DESIGN_HEATFLUX)
```
```suggestion
% Objective function in gradient evaluation
```

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+void CFlowOutput::Set_NearfieldInverseDesign(CSolver *solver, const CGeometry *geometry, const CConfig *config){
+
+  bool output = true;
+  ofstream EquivArea_file, FuncGrad_file;
+  unsigned short iMarker = 0, iDim;
+  su2double Gamma, auxXCoord, auxYCoord, auxZCoord, InverseDesign = 0.0, DeltaX, Coord_i, Coord_j, jp1Coord, *Coord = nullptr, MeanFuntion,
+  *Face_Normal = nullptr, auxArea, auxPress, Mach, Beta, R_Plane, Pressure_Inf,
+  ModVelocity_Inf, Velocity_Inf[3], factor,
+  jFunction, jp1Function;
+  unsigned long jVertex, iVertex, iPoint, nVertex_NearField = 0, auxPoint,
+  auxDomain;
+  unsigned short iPhiAngle;
+  ofstream NearFieldEA_file; ifstream TargetEA_file;
```
For all these variables please go through them and check whether it makes sense to declare them when they are used. E.g.
```c++
su2double ..., Mach, ... ;
Mach = config->GetMach();
```
to 
```c++
const su2double Mach = config->GetMach();
```
Also `const` them whenever possible (e.g. `nDim`). And don't worry about declaring a variable not at the top, that is totally ok (and even better imo) to keep the vars close to where they are needed

Please do that for all loop variables e.g. iDim, iPoint, iVertex unless you specificially need the var outside of its loop scope.
```c++
for(unsigned short iDim = 0; iDim < nDim; iDim++)
```

And feel free to use `auto` when you directly set a variable using an e.g. `Get<Value>` function e.g.
```c++
const auto nDim = geometry->GetnDim();
```
But do not use `auto` for anything that would be an `su2double`! (I know that might sound strange but there is an explanation to which I could give upon request)

pcarruscag replied:
I think we could be a little bit less demanding with a first-time contributor that is already cutting a chunk out of the output_legacy :smile:

TobiKattmann replied:
Yeah, I think it is ok to not do all that. On the other hand, this is the easier and rewarding cleanup part imo :D (especially compared to non blocking MPI comms... Personally I would prefer to procrastinate with cleaning variables every day of the week 🤫 )

snow54 replied:
The issue related to non blocking MPI comms either causes an error or stops execution unless I use a certain number of cores, so it has a higher priority for me. I'll address the rest of comments once I resolve such roadblocks.

snow54 replied:
I have changed declaration for easy ones. I'm not sure which ones make sense to be declared when they are used. For iDim, iPoint, and iVertex, they are used in multiple "for" loops. I'm not sure if they should be declared in multiple "for" loops or declared only once at the top.

pcarruscag replied:
Only for completeness and future developments (not something you have to address now IMO), these are the guidelines we should try to follow regarding variables declaration and initialization: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-introduce

snow54 replied:
Thank you for the reference. I will work on it in the future.

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  Velocity_Inf[1] = config->GetVelocity_FreeStreamND()[1];
+  Velocity_Inf[2] = config->GetVelocity_FreeStreamND()[2];
+  ModVelocity_Inf = 0;
+  for (iDim = 0; iDim < 3; iDim++)
```
either use `nDim` or `MAXNDIM` if it needs to stay 3

snow54 replied:
@pcarruscag has deleted those lines [here](https://github.com/su2code/SU2/pull/1329#discussion_r684619057).

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  if (rank == MASTER_NODE) cout << "Writing Equivalent Area files." << endl ;
+
+
+  int nProcessor = size;
```
Just use size for all nProcessor instances and get rid of that variable?

snow54 replied:
I have replaced nProcessor with size.

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+    /*--- Create a file with the equivalent area distribution at each azimuthal angle ---*/
+
+    NearFieldEA_file.precision(15);
```
There is a `config->GetOutput_Precision()` which could be used here. and then one can set OUTPUT_PRECISION=15 e.g. in the cfg

snow54 replied:
I have replaced 15 with `config->GetOutput_Precision()`.

TobiKattmann commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_adjoint.cfg:
```diff
@@ -0,0 +1,153 @@
+% Physical governing equations (EULER, NAVIER_STOKES)
+SOLVER= EULER
+%
+% Mathematical problem (DIRECT, CONTINUOUS_ADJOINT, DISCRETE_ADJOINT)
+MATH_PROBLEM= DISCRETE_ADJOINT
```
just as MATH_PROBLEM=DIRECT this line is not necessary and the Discrete adjoint mode is automatically used when you call the file using SU2_CFD_AD
Edit: Just as pedro mentioned, if this option is dividing the two cfg's then you can throw this option out and use the same cfg for both: direct and adjoint

TobiKattmann commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206_adjoint.cfg:
```diff
@@ -0,0 +1,153 @@
+%
+CFL_NUMBER=3
+CFL_ADAPT= NO
+%INNER_ITER= 4000
```
The testcases have the ability to set the iteration with the .test_iter option, i.e. you can keep the iteration count that is necessary for good convergence as the active option in the config. But it seems INNER_ITER and ITER can be used equivalently for steady-state cases. I.e. for the .test_iter option to work you might need to use ITER...
I hope that was somewhat understandable

snow54 replied:
Should I set the number of iterations so that I can confirm the converged solution matches expectation? I saw most of the test cases with 10 or 20 iterations (which I assume is not enough for good convergence), so I thought you just want to make sure that a interim result matches expectation.

pcarruscag replied:
That is correct. In your case, because you have a reference solution for the adjoint, you can even restart the direct simulation and run only 2 iterations.

snow54 replied:
I have changed INNER_ITER to ITER.

TobiKattmann commented on TestCases/parallel_regression.py:
```diff
@@ -185,6 +185,17 @@ def main():
+    ea_naca64206           = TestCase('ea_naca64206')
+    ea_naca64206.cfg_dir   = "optimization_euler/equivalentarea_naca64206"
+    ea_naca64206.cfg_file  = "NACA64206_direct.cfg"
+    ea_naca64206.test_iter = 20
```
this is what I meant above, use ITER= 4000 in the cfg and `ea_naca64206.test_iter = 40` here. 
In case you want to comment out options because they serve a purpose for easy switching between things please add an explaining comment why that is done and what expected changes are

snow54 replied:
See [this comment](https://github.com/su2code/SU2/pull/1329#discussion_r684608624).

TobiKattmann commented on TestCases/parallel_regression.py:
```diff
@@ -185,6 +185,17 @@ def main():
+    ea_naca64206.cfg_dir   = "optimization_euler/equivalentarea_naca64206"
+    ea_naca64206.cfg_file  = "NACA64206_direct.cfg"
+    ea_naca64206.test_iter = 20
+    ea_naca64206.test_vals = [-1.568224, 1.6373e-08]
```
More .test_vals i.e. more sreen_output for better overall coverage. Please use SU2_CFD -d <your.cfg> see available options and pack in a few more things that you feel comfortable with and think their respective values are reasonable (I am not exactly knowledgable when it comes to supersonic stuff)

snow54 replied:
I have increased the number of variables in `.test_vals` to 4.

TobiKattmann reviewed (commented):
Hi Yuki, Thanks for this contribution 💐 I have a few comments in the review. You didn't write the code initially but it might still be good idea to make the code a bit nicer overall. In case there are things that are not understandable please let me know.

I would highly appreciate if you add a little `README.md` to your testcase folder and explain a bit what that testcase is meant to showcase and what one is meant to see/look out for/expected outcome and just a hint how to use it (if that is explained in the cfg then just mention that). This will help down the road for anyone that wonders or has to fix a failing reg test and wonders what this is all about. You might also mention the PR number 1329 in there. Doesn't need to be a long text.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+  SU2_MPI::Allreduce(&nLocalVertex_NearField, &nVertex_NearField, 1, MPI_UNSIGNED_LONG, MPI_SUM, SU2_MPI::GetComm());
+  SU2_MPI::Allreduce(&nLocalVertex_NearField, &MaxLocalVertex_NearField, 1, MPI_UNSIGNED_LONG, MPI_MAX, SU2_MPI::GetComm());
+  SU2_MPI::Gather(Buffer_Send_nVertex, 1, MPI_UNSIGNED_LONG, Buffer_Receive_nVertex.data(), 1, MPI_UNSIGNED_LONG, MASTER_NODE, SU2_MPI::GetComm());
```
```suggestion
  SU2_MPI::Gather(&nLocalVertex_NearField, 1, MPI_UNSIGNED_LONG, Buffer_Receive_nVertex.data(), 1, MPI_UNSIGNED_LONG, MASTER_NODE, SU2_MPI::GetComm());
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+            nLocalVertex_NearField ++;
+      }
+
+  unsigned long *Buffer_Send_nVertex = new unsigned long [1];
+  Buffer_Send_nVertex[0] = nLocalVertex_NearField;
+
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  SU2_MPI::Allreduce(&nLocalVertex_NearField, &nVertex_NearField, 1, MPI_UNSIGNED_LONG, MPI_SUM, SU2_MPI::GetComm());
+  SU2_MPI::Allreduce(&nLocalVertex_NearField, &MaxLocalVertex_NearField, 1, MPI_UNSIGNED_LONG, MPI_MAX, SU2_MPI::GetComm());
+  SU2_MPI::Gather(Buffer_Send_nVertex, 1, MPI_UNSIGNED_LONG, Buffer_Receive_nVertex.data(), 1, MPI_UNSIGNED_LONG, MASTER_NODE, SU2_MPI::GetComm());
+  delete [] Buffer_Send_nVertex;
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  SU2_MPI::Gather(Buffer_Send_nVertex, 1, MPI_UNSIGNED_LONG, Buffer_Receive_nVertex.data(), 1, MPI_UNSIGNED_LONG, MASTER_NODE, SU2_MPI::GetComm());
+  delete [] Buffer_Send_nVertex;
+
+  vector<su2double> Buffer_Send_Xcoord          (MaxLocalVertex_NearField);
+  vector<su2double> Buffer_Send_Ycoord          (MaxLocalVertex_NearField);
+  vector<su2double> Buffer_Send_Zcoord          (MaxLocalVertex_NearField);
+  vector<unsigned long> Buffer_Send_IdPoint     (MaxLocalVertex_NearField);
+  vector<su2double> Buffer_Send_Pressure        (MaxLocalVertex_NearField);
+  vector<su2double> Buffer_Send_FaceArea        (MaxLocalVertex_NearField);
```
```suggestion
  vector<su2double> Buffer_Send_Xcoord          (MaxLocalVertex_NearField, 0.0);
  vector<su2double> Buffer_Send_Ycoord          (MaxLocalVertex_NearField, 0.0);
  vector<su2double> Buffer_Send_Zcoord          (MaxLocalVertex_NearField, 0.0);
  vector<unsigned long> Buffer_Send_IdPoint     (MaxLocalVertex_NearField, 0);
  vector<su2double> Buffer_Send_Pressure        (MaxLocalVertex_NearField, 0.0);
  vector<su2double> Buffer_Send_FaceArea        (MaxLocalVertex_NearField, 0.0);
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  unsigned long nBuffer_Pressure = MaxLocalVertex_NearField;
+  unsigned long nBuffer_FaceArea = MaxLocalVertex_NearField;
+
+  for (iVertex = 0; iVertex < MaxLocalVertex_NearField; iVertex++) {
+    Buffer_Send_IdPoint[iVertex] = 0; Buffer_Send_Pressure[iVertex] = 0.0;
+    Buffer_Send_FaceArea[iVertex] = 0.0; Buffer_Send_Xcoord[iVertex] = 0.0;
+    Buffer_Send_Ycoord[iVertex] = 0.0; Buffer_Send_Zcoord[iVertex] = 0.0;
+  }
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  SU2_MPI::Gather(Buffer_Send_Pressure.data(), nBuffer_Pressure, MPI_DOUBLE, Buffer_Receive_Pressure.data(), nBuffer_Pressure, MPI_DOUBLE, MASTER_NODE, SU2_MPI::GetComm());
+  SU2_MPI::Gather(Buffer_Send_FaceArea.data(), nBuffer_FaceArea, MPI_DOUBLE, Buffer_Receive_FaceArea.data(), nBuffer_FaceArea, MPI_DOUBLE, MASTER_NODE, SU2_MPI::GetComm());
+
+  vector<su2double> Xcoord;
+  vector<su2double> Ycoord;
+  vector<su2double> Zcoord;
+  vector<short> AzimuthalAngle;
+  vector<unsigned long> IdPoint;
+  vector<unsigned long> IdDomain;
+  vector<su2double> Pressure;
+  vector<su2double> FaceArea;
+  vector<su2double> EquivArea;
+  vector<su2double> TargetArea;
+  vector<su2double> NearFieldWeight;
+  vector<su2double> Weight;
+
+  if (rank == MASTER_NODE) {
+
+    Xcoord.resize(nVertex_NearField);
+    Ycoord.resize(nVertex_NearField);
+    Zcoord.resize(nVertex_NearField);
+    AzimuthalAngle.resize(nVertex_NearField);
+    IdPoint.resize(nVertex_NearField);
+    IdDomain.resize(nVertex_NearField);
+    Pressure.resize(nVertex_NearField);
+    FaceArea.resize(nVertex_NearField);
+    EquivArea.resize(nVertex_NearField);
+    TargetArea.resize(nVertex_NearField);
+    NearFieldWeight.resize(nVertex_NearField);
+    Weight.resize(nVertex_NearField);
```
```suggestion
  if (rank == MASTER_NODE) {

    vector<su2double> Xcoord(nVertex_NearField);
    vector<su2double> Ycoord(nVertex_NearField);
    vector<su2double> Zcoord(nVertex_NearField);
    vector<short> AzimuthalAngle(nVertex_NearField);
    vector<unsigned long> IdPoint(nVertex_NearField);
    vector<unsigned long> IdDomain(nVertex_NearField);
    vector<su2double> Pressure(nVertex_NearField);
    vector<su2double> FaceArea(nVertex_NearField);
    vector<su2double> EquivArea(nVertex_NearField);
    vector<su2double> TargetArea(nVertex_NearField);
    vector<su2double> NearFieldWeight(nVertex_NearField);
    vector<su2double> Weight(nVertex_NearField);
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+      }
+
+  }
+
+
+  if (rank == MASTER_NODE) {
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+    /*--- Create vectors and distribute the values among the different PhiAngle queues ---*/
+
+    vector<vector<su2double> > Xcoord_PhiAngle; Xcoord_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > Ycoord_PhiAngle; Ycoord_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > Zcoord_PhiAngle; Zcoord_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<unsigned long> > IdPoint_PhiAngle; IdPoint_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<unsigned long> > IdDomain_PhiAngle; IdDomain_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > Pressure_PhiAngle; Pressure_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > FaceArea_PhiAngle; FaceArea_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > EquivArea_PhiAngle; EquivArea_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > TargetArea_PhiAngle; TargetArea_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > NearFieldWeight_PhiAngle; NearFieldWeight_PhiAngle.resize(PhiAngleList.size());
+    vector<vector<su2double> > Weight_PhiAngle; Weight_PhiAngle.resize(PhiAngleList.size());
```
```suggestion
    vector<vector<su2double> > Xcoord_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > Ycoord_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > Zcoord_PhiAngle(PhiAngleList.size());
    vector<vector<unsigned long> > IdPoint_PhiAngle(PhiAngleList.size());
    vector<vector<unsigned long> > IdDomain_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > Pressure_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > FaceArea_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > EquivArea_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > TargetArea_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > NearFieldWeight_PhiAngle(PhiAngleList.size());
    vector<vector<su2double> > Weight_PhiAngle(PhiAngleList.size());
```

snow54 commented:
Hi @TobiKattmann, thanks for your review. I'll try to do what I can within this week as I don't have much time from next week.

> Hi Yuki, Thanks for this contribution bouquet I have a few comments in the review. You didn't write the code initially but it might still be good idea to make the code a bit nicer overall. In case there are things that are not understandable please let me know.

As for things that I don't understand, see below although I would like to separate them from this PR.
- WeightNF.dat
  - I don't understand what this is for. It is read in `CAdjEulerSolver::SetIntBoundary_Jump`.
- VENKATAKRISHNAN limiter
  - With MARKER_NEARFIELD defined, this limiter causes strange pressure disturbance near the nearfield boundary. (That's why I'm using VAN_ALBADA_EDGE in the test cases)
- `Bad match for point 47742.	Nearest donor distance: 3.6066950142e+00.`
  - Lots of this message shows up. I don't know what this message is for. It doesn't seem to cause any real issues, so I've been ignoring them.

TobiKattmann commented:
Hi 
> WeightNF.dat: I don't understand what this is for. It is read in CAdjEulerSolver::SetIntBoundary_Jump to calculate derivative, so it is used for adjoint calculation? My understanding is that at least discrete adjoint does not need to calculate derivative. Is it for continuous adjoint?

As far as I see it that function should not be called with the discrete adjoint solver. `CAdjEulerSolver::SetIntBoundary_Jump ` is only called by `CAdjFluidIteration::Preprocess` which should happen during a discrete adjoint run (but as far as i understand you that is not happening currently). All `CAdj<sth>` things are related to the continuous adjoint and `CDiscAdj<sth>` to the discrete adjoint.
Boundary conditions (this one being internal) need to be explicitly derived and set in the continuous adjoint but not in the discrete adjoint so your understanding was going in the right direction. (Although I am no expert in cont.adj. things either :) )

> VENKATAKRISHNAN limiter: With MARKER_NEARFIELD defined, this limiter causes strange pressure disturbance near the nearfield boundary. I tried to find a cause but no luck so far. (That's why I'm using VAN_ALBADA_EDGE in the test cases.)

you can try maybe to tune the `VENKAT_LIMITER_COEFF` in the cfg, but I have no better hint for now

> Bad match for point 47742. Nearest donor distance: 3.6066950142e+00.: Lots of this message shows up. I don't know what this message is for. It doesn't seem to cause any real issues, so I've been ignoring them.

That is certainly sth I would not ignore :) Does this occur with the case you added (i.e. no periodic bc)? Does it also occur when you run single core? maybe it is matching between internal MPI boundaries. In that case yould can add RANK to the VOLUME_OUTPUT and take more and more cores and see when it fails -> look at the Rank boundaries and maybe you notice sth

snow54 commented:
> > WeightNF.dat: I don't understand what this is for. It is read in CAdjEulerSolver::SetIntBoundary_Jump to calculate derivative, so it is used for adjoint calculation? My understanding is that at least discrete adjoint does not need to calculate derivative. Is it for continuous adjoint?
> 
> As far as I see it that function should not be called with the discrete adjoint solver. `CAdjEulerSolver::SetIntBoundary_Jump ` is only called by `CAdjFluidIteration::Preprocess` which should happen during a discrete adjoint run (but as far as i understand you that is not happening currently). All `CAdj<sth>` things are related to the continuous adjoint and `CDiscAdj<sth>` to the discrete adjoint.
> Boundary conditions (this one being internal) need to be explicitly derived and set in the continuous adjoint but not in the discrete adjoint so your understanding was going in the right direction. (Although I am no expert in cont.adj. things either :) )

So, it sounds like I don't need to care about it as long as I use discrete adjoint then.

> > VENKATAKRISHNAN limiter: With MARKER_NEARFIELD defined, this limiter causes strange pressure disturbance near the nearfield boundary. I tried to find a cause but no luck so far. (That's why I'm using VAN_ALBADA_EDGE in the test cases.)
> 
> you can try maybe to tune the `VENKAT_LIMITER_COEFF` in the cfg, but I have no better hint for now

I forgot to mention. I have tested VENKATAKRISHNAN limiter with v6. I am hoping that it is fixed by now after reading [this discussion](https://github.com/su2code/SU2/issues/917).

> > Bad match for point 47742. Nearest donor distance: 3.6066950142e+00.: Lots of this message shows up. I don't know what this message is for. It doesn't seem to cause any real issues, so I've been ignoring them.
> 
> That is certainly sth I would not ignore :) Does this occur with the case you added (i.e. no periodic bc)? Does it also occur when you run single core? maybe it is matching between internal MPI boundaries. In that case yould can add RANK to the VOLUME_OUTPUT and take more and more cores and see when it fails -> look at the Rank boundaries and maybe you notice sth

Yes, it occurs with the case I added. I haven't tested with single core. As far as I see in `CPhysicalGeometry::MatchNearField`, the code looks for a nearest point to a point on the nearfield boundary. If it exceeds 1e-1 (which is usually the case if you have real scale mesh), it prints this warning. I'm not sure what this is for. I'm ignoring it since I don't see an issue with this distance exceeding 1e-1.

snow54 commented:
@TobiKattmann 
I have added 'READ.md' to the testcase folder. Let me know what you think. I'm not exactly sure what I should write as expected outcome for discrete adjoint solver. Ultimately, it is used for shape optimization, so getting a better objective function is what I need to confirm but I cannot run a complete shape optimization in the test case with only a few iterations. What do you usually look for to confirm adjoint calculation is working as expected?

> I would highly appreciate if you add a little `README.md` to your testcase folder and explain a bit what that testcase is meant to showcase and what one is meant to see/look out for/expected outcome and just a hint how to use it (if that is explained in the cfg then just mention that). This will help down the road for anyone that wonders or has to fix a failing reg test and wonders what this is all about. You might also mention the PR number 1329 in there. Doesn't need to be a long text.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,504 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  Beta = sqrt(Mach*Mach-1.0);
+  R_Plane = fabs(config->GetEA_IntLimit(2));
+  Pressure_Inf = config->GetPressure_FreeStreamND();
+  Velocity_Inf[0] = config->GetVelocity_FreeStreamND()[0];
+  Velocity_Inf[1] = config->GetVelocity_FreeStreamND()[1];
+  Velocity_Inf[2] = config->GetVelocity_FreeStreamND()[2];
+  ModVelocity_Inf = 0;
+  for (iDim = 0; iDim < 3; iDim++)
+    ModVelocity_Inf += Velocity_Inf[iDim] * Velocity_Inf[iDim];
```
```suggestion
  const su2double ModVelocity_Inf = GeometryToolbox::SquaredNorm(3, config->GetVelocity_FreeStreamND());
```
You may need an #include geometry_toolbox.hpp

snow54 replied:
geometry_toolbox.hpp is already included.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,504 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  ofstream EquivArea_file, FuncGrad_file;
+  unsigned short iMarker = 0, iDim;
+  su2double Gamma, auxXCoord, auxYCoord, auxZCoord, InverseDesign = 0.0, DeltaX, Coord_i, Coord_j, jp1Coord, *Coord = nullptr, MeanFuntion,
+  *Face_Normal = nullptr, auxArea, auxPress, Mach, Beta, R_Plane, Pressure_Inf,
+  ModVelocity_Inf, Velocity_Inf[3], factor,
```
```suggestion
  *Face_Normal = nullptr, auxArea, auxPress, Mach, Beta, R_Plane, Pressure_Inf, factor,
```

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -828,63 +827,58 @@ void CEulerSolver::Set_MPI_Nearfield(CGeometry *geometry, CConfig *config) {
 
   }
-
+  
```
```suggestion

```

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -828,63 +827,58 @@ void CEulerSolver::Set_MPI_Nearfield(CGeometry *geometry, CConfig *config) {
+  
   /*--- Allocate the buffer vectors in the appropiate domain (master, iDomain) ---*/
-
+  
```
```suggestion

```

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -828,63 +827,58 @@ void CEulerSolver::Set_MPI_Nearfield(CGeometry *geometry, CConfig *config) {
-  SU2_MPI::Barrier(SU2_MPI::GetComm());
-
-  /*--- Initialize the counters for the larger send buffers (by domain) ---*/
+  SU2_MPI::Waitall(size-1,&req_Recv[0],MPI_STATUS_IGNORE);
```
```suggestion
  SU2_MPI::Waitall(size-1,req_Recv,MPI_STATUS_IGNORE);
```

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -973,46 +1019,36 @@ void CEulerSolver::Set_MPI_Nearfield(CGeometry *geometry, CConfig *config) {
   /*--- Wait for the non-blocking sends to complete. ---*/
 
-  SU2_MPI::Barrier(SU2_MPI::GetComm());
+  SU2_MPI::Waitall(size-1,&req_Recv[0],MPI_STATUS_IGNORE);
```
```suggestion
  SU2_MPI::Waitall(size-1,req_Recv,MPI_STATUS_IGNORE);
```

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,144 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%                                                                              %
+% SU2 configuration file                                                       %
+% Case description: NACA64-206 in inviscid, supersonic flow                  %
+% Author: Yuki Utsumi	      	      	              	       	       %
+% Institution: Individual				               %
+% Date: 2021.08.07                                                            %
+% File Version 7.1.1 "Blackbird"                                                %
+%                                                                              %
```
Do you want to make this all pretty? :)

snow54 replied:
Not really. I just tried to mimic some of existing config files in TestCases. I have deleted those lines.

pcarruscag replied:
Hmm ok I was just pointing out that the "%" were not aligned, which can happen if you used tabs.

snow54 replied:
Ah, I misunderstood. I brought back those lines and aligned "%".

pcarruscag commented on TestCases/optimization_euler/equivalentarea_naca64206/README.md:
```diff
@@ -0,0 +1,39 @@
+# Test case for Equivalent area calculation 
```
Awesome :+1:

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -764,31 +764,30 @@ void CEulerSolver::Set_MPI_ActDisk(CSolver **solver_container, CGeometry *geomet
+  unsigned long *nPointTotal_s       = new unsigned long[size];
+  unsigned long *nPointTotal_r       = new unsigned long[size];
+  su2double     *iPrimVar            = new su2double [nPrimVar];
+  unsigned long nPointTotal_rCum[size+1];
```
You have to make this one a vector<unsigned long>, as variable-length arrays are not standard c++.

snow54 replied:
I had trouble using vectors, so I used arrays. I'll try to work on it tomorrow.

snow54 replied:
Actually, I was able to fix them easily.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,498 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+  bool output = true;
+  ofstream EquivArea_file, FuncGrad_file;
+  unsigned short iMarker = 0, iDim;
```
```suggestion
  unsigned short iMarker = 0;
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,498 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  Beta = sqrt(Mach*Mach-1.0);
+  R_Plane = fabs(config->GetEA_IntLimit(2));
+  Pressure_Inf = config->GetPressure_FreeStreamND();
+  const su2double ModVelocity_Inf = GeometryToolbox::SquaredNorm(3, config->GetVelocity_FreeStreamND());
```
```suggestion
```

pcarruscag commented on TestCases/parallel_regression.py:
```diff
@@ -185,6 +185,17 @@ def main():
+    ea_naca64206.cfg_dir   = "optimization_euler/equivalentarea_naca64206"
+    ea_naca64206.cfg_file  = "NACA64206.cfg"
+    ea_naca64206.test_iter = 10
+    ea_naca64206.test_vals = [-1.086302, -0.391816, -0.002329, 6.7775e+04]
```
```suggestion
    ea_naca64206.test_vals = [-1.076215, -0.391987, -0.000701, 67775.0]
```

pcarruscag commented on TestCases/parallel_regression_AD.py:
```diff
@@ -75,6 +75,17 @@ def main():
+    ea_naca64206.cfg_dir      = "optimization_euler/equivalentarea_naca64206"
+    ea_naca64206.cfg_file     = "NACA64206.cfg"
+    ea_naca64206.test_iter    = 10
+    ea_naca64206.test_vals    = [2.805558, 2.096837, 2.3722e+06, 4.5048e+00]
```
```suggestion
    ea_naca64206.test_vals    = [2.782467, 2.056357, 2372200.0, 3.878]
```

pcarruscag reviewed (commented):
It's normal for residuals to be different in the test environment than locally for you

pcarruscag commented:
Have you been able to converge the adjoint now that MPI problems are fixed, and verify the derivatives you obtain?

snow54 commented:
> Have you been able to converge the adjoint now that MPI problems are fixed, and verify the derivatives you obtain?

For convergence, yes. For verifying the derivatives, I'm looking for a way to do it now.

snow54 commented:
> > Have you been able to converge the adjoint now that MPI problems are fixed, and verify the derivatives you obtain?
> 
> For convergence, yes. For verifying the derivatives, I'm looking for a way to do it now.

It seems that I can use the following python scripts to get derivatives.

- `discrete_adjoint.py`
- `finite_differences.py`
- `direct_differentiation.py`

`discrete_adjoint.py` and `finite_differences.py` output similar values although they differ sometimes by 20% or so. Is such difference expected? `direct_differentiation.py` outputs do not match at all and I'm trying to understand the reason.

pcarruscag commented:
I would focus more on discrete adjoint Vs finite differences, the direct mode can be tricky sometimes.
About the discrepancy, you can try changing the finite difference step (especially if the 20% error is for small derivatives)

pcarruscag reviewed (approved):
Thank you again for reviving this feature.
I think this is good to go but let's give Tobi and others some time to also take a final look.

pcarruscag reviewed (commented):
Ah! You commited different version of codi and opdi submodules, you need to revert that.
Checkout develop
Git submodule update externals/codi and opdi
Checkout this branch
Commit -a
(Checkout does not update submodules)

snow54 commented:
> Ah! You commited different version of codi and opdi submodules, you need to revert that.
> Checkout develop
> Git submodule update externals/codi and opdi
> Checkout this branch
> Commit -a
> (Checkout does not update submodules)

Hmm, nothing happened by `git submodule update`. I did some research on the web and I did `git submodule foreach git pull origin master`. It updated medi, meson, ninja, mutationpp instead of codi and opdi. I'll try to fix it tomorrow.

snow54 commented:
> Thank you again for reviving this feature.
> I think this is good to go but let's give Tobi and others some time to also take a final look.

Thank you for giving me a lot of advice. I noticed the number of comments on this PR is much larger than others. I will still make some updates while waiting for others to review.

snow54 commented:
> I would focus more on discrete adjoint Vs finite differences, the direct mode can be tricky sometimes.
> About the discrepancy, you can try changing the finite difference step (especially if the 20% error is for small derivatives)

I have reduced FIN_DIFF_STEP from 0.001 to 0.0001. Discrepancy is about 2% now. Can I consider this is acceptable?

pcarruscag commented:
Sounds good for verification.
No codi and opdi are still not correct, the versions used in master are not the ones used now in develop.
Go into the folders of those submodules and do checkout 3c3211fef2e225ab89680a4063b62bb3bb38a7e4 for codi
and 2735b503f60163e8d64e1ac56cce46173a9fd4a9 for opdi
Then go back to the root of SU2 and commit.

snow54 commented:
> No codi and opdi are still not correct, the versions used in master are not the ones used now in develop.
> Go into the folders of those submodules and do checkout 3c3211fef2e225ab89680a4063b62bb3bb38a7e4 for codi
> and 2735b503f60163e8d64e1ac56cce46173a9fd4a9 for opdi
> Then go back to the root of SU2 and commit.

Thanks! No changes on codi and opdi is shown anymore.

snow54 commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -789,6 +789,530 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+void CFlowOutput::Set_NearfieldInverseDesign(CSolver *solver, const CGeometry *geometry, const CConfig *config){
+
+  bool output = true;
```
I have removed `output`. Before ver.7.0.0, this is called out in two different places, `CDiscAdjFluidDriver::SetObjFunction` and `COutput::SetConvHistory_Body`. `output` was used to determine if output files dedicated to equivalent area are generated. Currently, it is only called out once in `CFlowCompOutput::LoadHistoryData`, so I think it is OK to generate those output files anytime when this function is used.

snow54 commented:
Hi @TobiKattmann 

> > > Bad match for point 47742. Nearest donor distance: 3.6066950142e+00.: Lots of this message shows up. I don't know what this message is for. It doesn't seem to cause any real issues, so I've been ignoring them.
> > 
> > 
> > That is certainly sth I would not ignore :) Does this occur with the case you added (i.e. no periodic bc)? Does it also occur when you run single core? maybe it is matching between internal MPI boundaries. In that case yould can add RANK to the VOLUME_OUTPUT and take more and more cores and see when it fails -> look at the Rank boundaries and maybe you notice sth
> 
> Yes, it occurs with the case I added. I haven't tested with single core. As far as I see in `CPhysicalGeometry::MatchNearField`, the code looks for a nearest point to a point on the nearfield boundary. If it exceeds 1e-1 (which is usually the case if you have real scale mesh), it prints this warning. I'm not sure what this is for. I'm ignoring it since I don't see an issue with this distance exceeding 1e-1.

I tested with single core and this message appears as well. Again, I don't see a point of this message and need to know the original intent.

snow54 commented:
I think I have addressed all the comments. Let me know if there is anything that still needs to be fixed.
I have created a PR#69 in the TestCases repository to update FFD box in the mesh file. Please merge it. It is not required for regression but it is required if someone wants to check derivatives are correctly calculated.

As I mentioned in the previous post, there are still some remaining issues shown below. They were there before this PR. I would like to defer them to future PRs once I understand more about how SU2 works.

- WeightNF.dat
  - I don't understand what this is for. It is read in CAdjEulerSolver::SetIntBoundary_Jump to calculate derivative, so it is used for adjoint calculation? My understanding is that at least discrete adjoint does not need to calculate derivative. Is it for continuous adjoint?
- VENKATAKRISHNAN limiter
  - With MARKER_NEARFIELD defined, this limiter causes strange pressure disturbance near the nearfield boundary. I tried to find a cause but no luck so far. (That's why I'm using VAN_ALBADA_EDGE in the test cases.)
- `Bad match for point 47742. Nearest donor distance: 3.6066950142e+00.`
  - Lots of this message shows up. I don't know what this message is for. It doesn't seem to cause any real issues, so I've been ignoring them.

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+    else
+      EquivArea_file << "VARIABLES = \"Height (m) at r="<< R_Plane << " m. (cylindrical coordinate system)\",\"Equivalent Area (m<sup>2</sup>)\",\"Target Equivalent Area (m<sup>2</sup>)\",\"Cp\"" << "\n";
+
+    for (iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
+      EquivArea_file << fixed << "ZONE T= \"<greek>F</greek>=" << PhiAngleList[iPhiAngle] << " deg.\"" << "\n";
+      for (iVertex = 0; iVertex < Xcoord_PhiAngle[iPhiAngle].size(); iVertex++) {
```
```suggestion
    for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
      EquivArea_file << fixed << "ZONE T= \"<greek>F</greek>=" << PhiAngleList[iPhiAngle] << " deg.\"" << "\n";
      for (unsigned long iVertex = 0; iVertex < Xcoord_PhiAngle[iPhiAngle].size(); iVertex++) {
```

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+    FuncGrad_file.open("WeightNF.dat", ios::out);
+
+    FuncGrad_file << scientific << "-1.0";
+    for (iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
+      FuncGrad_file << scientific << "\t" << PhiAngleList[iPhiAngle];
+    FuncGrad_file << "\n";
+
+    for (iVertex = 0; iVertex < NearFieldWeight_PhiAngle[0].size(); iVertex++) {
+      su2double XcoordRot = Xcoord_PhiAngle[0][iVertex]*cos(AoA) - Zcoord_PhiAngle[0][iVertex]*sin(AoA);
+      FuncGrad_file << scientific << XcoordRot;
+      for (iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```
```suggestion
    for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
      FuncGrad_file << scientific << "\t" << PhiAngleList[iPhiAngle];
    FuncGrad_file << "\n";

    for (unsigned long iVertex = 0; iVertex < NearFieldWeight_PhiAngle[0].size(); iVertex++) {
      su2double XcoordRot = Xcoord_PhiAngle[0][iVertex]*cos(AoA) - Zcoord_PhiAngle[0][iVertex]*sin(AoA);
      FuncGrad_file << scientific << XcoordRot;
      for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+void CFlowOutput::Set_NearfieldInverseDesign(CSolver *solver, const CGeometry *geometry, const CConfig *config){
+
+  ofstream EquivArea_file, FuncGrad_file;
+  unsigned short iMarker = 0;
+  su2double auxXCoord, auxYCoord, auxZCoord, InverseDesign = 0.0, DeltaX,
+    Coord_i, Coord_j, jp1Coord, *Coord = nullptr, MeanFunction,
+    *Face_Normal = nullptr, auxArea, auxPress, jFunction, jp1Function;
+  unsigned long jVertex, iVertex, iPoint, nVertex_NearField = 0, auxPoint,
+    auxDomain;
+  unsigned short iPhiAngle;
```
```suggestion
  su2double auxXCoord, auxYCoord, auxZCoord, InverseDesign, DeltaX,
    Coord_i, Coord_j, jp1Coord, *Coord = nullptr, MeanFunction,
    *Face_Normal = nullptr, auxArea, auxPress, jFunction, jp1Function;
  unsigned long iPoint, auxPoint, auxDomain;
```

TobiKattmann replied:
I am somewhat certain that removing iMarker, i,jVertex should be fine with the below suggested changes. But I wouldnt bet on myself here.

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  su2double XCoordEnd_OF = config->GetEA_IntLimit(1);
+
+  unsigned short nDim = geometry->GetnDim();
+  su2double AoA = -(config->GetAoA()*PI_NUMBER/180.0);
+  su2double EAScaleFactor = config->GetEA_ScaleFactor(); // The EA Obj. Func. should be ~ force based Obj. Func.
+
+  const su2double Mach  = config->GetMach();
+  const su2double Gamma = config->GetGamma();
+  const su2double Beta = sqrt(Mach*Mach-1.0);
```
👍 on the lower ones. While you're at it, you might try to get others (like the upper ones) const as well if possible. Now that there is a working testcase

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+  if (rank == MASTER_NODE) cout << "Writing Equivalent Area files." << endl ;
+
+  unsigned long nLocalVertex_NearField = 0, MaxLocalVertex_NearField = 0;
+  int iProcessor;
+
```
```suggestion
```

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+  /*--- Compute the total number of points of the near-field ghost nodes ---*/
+
+  nLocalVertex_NearField = 0;
```
```suggestion
  unsigned long nLocalVertex_NearField = 0;
```

TobiKattmann commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+            nLocalVertex_NearField ++;
+      }
+
+  /*--- Send Near-Field vertex information --*/
+
```
```suggestion
  /*--- Send Near-Field vertex information --*/
  unsigned long MaxLocalVertex_NearField, nVertex_NearField;

```

TobiKattmann reviewed (approved):
Thanks @snow54 for tackling this 💐 . Especially for the effort on documenting the Testcase. 

towards my chunk of suggestions: I am well aware that you took the effort to move that stuff from the legacy output to the new, making it work and and add a testcase. You are not responsible to clean other peoples code. But that said, I think it makes sense to now spend a few extra minutes for a little more cleaning. That might spare you some headache down the road as well. Especially now that there is a testcase one can be a bit more aggressive on that ;)

I am not sure if maybe forgot to localize one the loop variables so it might fail on that. (Note that you can add the suggestions to a batch and make it one commit in the files-tab of the PR)

Otherwise it looks good to me

TobiKattmann commented:
>  Yes, it occurs with the case I added. I haven't tested with single core. As far as I see in CPhysicalGeometry::MatchNearField,  
 the code looks for a nearest point to a point on the nearfield boundary. If it exceeds 1e-1 (which is usually the case if you have 
 real scale mesh), it prints this warning. I'm not sure what this is for. I'm ignoring it since I don't see an issue with this distance 
 exceeding 1e-1.

Concerning this: I understand this nearfield marker as an internal "boundary" which does not alter the usual solution process at all. It is just added to compute some interesting quantity one. So the Nearfield marker needs to be in the mesh right, at the end like all other boundary markers? As such it is defined using mesh points and therefore there should be no discrepancies between nearfield marker nodes and mesh nodes? Am i correct with that, if not please enlighten me what is actually happening :)

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,483 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  su2double XCoordBegin_OF = config->GetEA_IntLimit(0);
+  su2double XCoordEnd_OF = config->GetEA_IntLimit(1);
+
+  unsigned short nDim = geometry->GetnDim();
```
already a member variable, can be removed

pcarruscag commented:
If that is what it is supposed to do we might have problems :grimacing:

pcarruscag commented:
For example Green-Gauss gradients have an exclusion for internal boundaries but not for nearfield ones.

TobiKattmann commented on TestCases/optimization_euler/equivalentarea_naca64206/NACA64206.cfg:
```diff
@@ -0,0 +1,180 @@
+OPT_BOUND_UPPER= 2
+OPT_BOUND_LOWER= -2
+FIN_DIFF_STEP= 0.0001
+DEFINITION_DV= ( 11, 1.0 | Wing | WING, 0, 0, 0, 0.0, 0.0, 1.0 ); ( 11, 1.0 | Wing | WING, 1, 0, 0, 0.0, 0.0, 1.0 ); ( 11, 1.0 | Wing | WING, 2, 0, 0, 0.0, 0.0, 1.0 ); ( 11, 1.0 | Wing | WING, 3, 0, 0, 0.0, 0.0, 1.0 ); ( 11, 1.0 | Wing | WING, 4, 0, 0, 0.0, 0.0, 1.0 ); ( 11, 1.0 | Wing | WING, 5, 0, 0, 0.0, 0.0, 1.0 ); ( 11, 1.0 | Wing | WING, 6, 0, 0, 0.0, 0.0, 1.0 ); ( 11, 1.0 | Wing | WING, 7, ...
```
Wait, it is just a single '\'

snow54 replied:
I'll fix this. `gedit` on Ubuntu had difficulty opening this super long line, so it will make it easy to handle this.
This is a copy-and-paste from `set_ffd_design_var.py`, so it might be better to edit this python script so it provides outputs that are easier to handle.

snow54 commented:
Hi @TobiKattmann 

Thank you for making lots of suggestions. I'm out of town now, so I'll address those once I get back later this week.

> > Yes, it occurs with the case I added. I haven't tested with single core. As far as I see in CPhysicalGeometry::MatchNearField,
> > the code looks for a nearest point to a point on the nearfield boundary. If it exceeds 1e-1 (which is usually the case if you have
> > real scale mesh), it prints this warning. I'm not sure what this is for. I'm ignoring it since I don't see an issue with this distance
> > exceeding 1e-1.
> 
> Concerning this: I understand this nearfield marker as an internal "boundary" which does not alter the usual solution process at all. It is just added to compute some interesting quantity one. So the Nearfield marker needs to be in the mesh right, at the end like all other boundary markers? As such it is defined using mesh points and therefore there should be no discrepancies between nearfield marker nodes and mesh nodes? Am i correct with that, if not please enlighten me what is actually happening :)

Yes, nearfield marker is just used to extract required physical quantities and should not alter the usual solution process at all. There should be no discrepancies between nearfield marker nodes and mesh nodes. I haven't checked this part of the code for a while but if I remember correctly, the "Bad match..." message is shown if there is discrepancy between a nearfield marker node and the nearest mesh node that is NOT on nearfield boundary.

> For example Green-Gauss gradients have an exclusion for internal boundaries but not for nearfield ones.

@pcarruscag So, I need to make something similar to internal boundaries? The flowfield looked reasonable except when I use VENKATAKRISHNAN limiter though where the nearfield boundary seemed to act like half permeable wall.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  /*--- Compute the total number of points of the near-field ghost nodes ---*/
+
+  unsigned long nLocalVertex_NearField = 0;
+  for (unsignde short iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++)
```
```suggestion
  for (unsigned short iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++)
```

TobiKattmann replied:
dank u well

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+  unsigned long iPoint, auxPoint, auxDomain;
+  ofstream NearFieldEA_file; ifstream TargetEA_file;
+
+  su2double XCoordBegin_OF = config->GetEA_IntLimit(0);
+  su2double XCoordEnd_OF = config->GetEA_IntLimit(1);
+
+  su2double AoA = -(config->GetAoA()*PI_NUMBER/180.0);
+  su2double EAScaleFactor = config->GetEA_ScaleFactor(); // The EA Obj. Func. should be ~ force based Obj. Func.
```
```suggestion
  const su2double XCoordBegin_OF = config->GetEA_IntLimit(0);
  const su2double XCoordEnd_OF = config->GetEA_IntLimit(1);

  const su2double AoA = -(config->GetAoA()*PI_NUMBER/180.0);
  const su2double EAScaleFactor = config->GetEA_ScaleFactor(); // The EA Obj. Func. should be ~ force based Obj. Func.
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+    /*--- Order the arrays (x Coordinate, Pressure, Point, and Domain) ---*/
+
+    for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```
```suggestion
    for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+    /*--- Check that all the azimuth lists have the same size ---*/
+
+    auto nVertex = Xcoord_PhiAngle[0].size();
+    for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```
```suggestion
    for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+    /*--- Compute equivalent area distribution at each azimuth angle ---*/
+
+    for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```
```suggestion
    for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+    if (TargetEA_file.fail()) {
+      /*--- Set the table to 0 ---*/
+      for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```
```suggestion
      for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+        TargetArea_PhiAngle_Trans.push_back(row);
+      }
+
+      for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```
```suggestion
      for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+    /*--- Evaluate the objective function ---*/
+
+    InverseDesign = 0;
+    for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```
```suggestion
    for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++)
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+            Buffer_Send_IdPoint[nLocalVertex_NearField] = iPoint;
+            Buffer_Send_Xcoord[nLocalVertex_NearField] = geometry->nodes->GetCoord(iPoint, 0);
+            Buffer_Send_Ycoord[nLocalVertex_NearField] = geometry->nodes->GetCoord(iPoint, 1);
+            Buffer_Send_Zcoord[nLocalVertex_NearField] = geometry->nodes->GetCoord(iPoint, 2);
```
```suggestion
            if (nDim == 3) {
              Buffer_Send_Zcoord[nLocalVertex_NearField] = geometry->nodes->GetCoord(iPoint, 2);
            }
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+          /*--- Compute the Azimuthal angle ---*/
+
+          su2double AngleDouble; short AngleInt;
+          AngleDouble = fabs(atan(-YcoordRot/ZcoordRot)*180.0/PI_NUMBER);
```
```suggestion
          const su2double AngleDouble = fabs(atan(-YcoordRot/ZcoordRot)*180.0/PI_NUMBER);
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+          if ((AngleDouble >= FixAzimuthalLine - 0.1) && (AngleDouble <= FixAzimuthalLine + 0.1))
+            AngleDouble = FixAzimuthalLine - 0.1;
+
+          AngleInt = SU2_TYPE::Short(floor(AngleDouble + 0.5));
```
```suggestion
          const auto AngleInt = SU2_TYPE::Short(floor(AngleDouble + 0.5));
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+    vector<su2double> Weight(nVertex_NearField);
+
+    nVertex_NearField = 0;
+    for (int iProcessor = 0; iProcessor < size; iProcessor++)
```
```suggestion
    for (int iProcessor = 0; iProcessor < size; iProcessor++) {
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,477 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+          IdDomain[nVertex_NearField] = iProcessor;
+          nVertex_NearField++;
+        }
+
+      }
```
```suggestion
      }
    }
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,478 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+
+          /*--- Compute the Azimuthal angle ---*/
+
+          const su2double AngleDouble = fabs(atan(-YcoordRot/ZcoordRot)*180.0/PI_NUMBER);
```
```suggestion
          su2double AngleDouble = fabs(atan(-YcoordRot/ZcoordRot)*180.0/PI_NUMBER);
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,478 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+void CFlowOutput::Set_NearfieldInverseDesign(CSolver *solver, const CGeometry *geometry, const CConfig *config){
+
+  ofstream EquivArea_file, FuncGrad_file;
+  su2double auxXCoord, auxYCoord, auxZCoord, InverseDesign, DeltaX,
```
```suggestion
  su2double auxXCoord, auxYCoord, auxZCoord, InverseDesign = 0.0, DeltaX,
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,478 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+    else
+      NearFieldEA_file << "VARIABLES = \"Height (m) at r="<< R_Plane << " m. (cylindrical coordinate system)\"";
+
+    for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```
```suggestion
    for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -784,6 +784,478 @@ void CFlowOutput::Set_CpInverseDesign(CSolver *solver, const CGeometry *geometry
+      else
+        NearFieldEA_file << scientific << (XcoordRot - XcoordRot_init);
+
+      for (unsigned short iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```
```suggestion
      for (unsigned long iPhiAngle = 0; iPhiAngle < PhiAngleList.size(); iPhiAngle++) {
```

TobiKattmann commented:
👍 This PR could be the one with the highest conversation count. very gud

TobiKattmann commented:
@pcarruscag I guess the last to changes were in the equivalent area computation routine so they shouldn't alter the result of 'discadj_incomp_cylinder' in hybrid_regression_AD? Why does it fail?

snow54 commented:
Thanks for merging this PR!

snow54 commented:
> Hi @TobiKattmann
> 
> Thank you for making lots of suggestions. I'm out of town now, so I'll address those once I get back later this week.
> 
> > > Yes, it occurs with the case I added. I haven't tested with single core. As far as I see in CPhysicalGeometry::MatchNearField,
> > > the code looks for a nearest point to a point on the nearfield boundary. If it exceeds 1e-1 (which is usually the case if you have
> > > real scale mesh), it prints this warning. I'm not sure what this is for. I'm ignoring it since I don't see an issue with this distance
> > > exceeding 1e-1.
> > 
> > 
> > Concerning this: I understand this nearfield marker as an internal "boundary" which does not alter the usual solution process at all. It is just added to compute some interesting quantity one. So the Nearfield marker needs to be in the mesh right, at the end like all other boundary markers? As such it is defined using mesh points and therefore there should be no discrepancies between nearfield marker nodes and mesh nodes? Am i correct with that, if not please enlighten me what is actually happening :)
> 
> Yes, nearfield marker is just used to extract required physical quantities and should not alter the usual solution process at all. There should be no discrepancies between nearfield marker nodes and mesh nodes. I haven't checked this part of the code for a while but if I remember correctly, the "Bad match..." message is shown if there is discrepancy between a nearfield marker node and the nearest mesh node that is NOT on nearfield boundary.

I did more investigations. This "Bad match..." message is output in `CPhysicalGeometry::MatchNearField` and this function has `SetDonorPoint`. However, currently there is no `GetDonorPoint` except for periodic boundary condition. I don't believe this is used any more. I went back in history and found out that `GetDonorPoint` exists in `CGridAdaptation::SetHomothetic_Adaptation2D` in ver.6.2.0. I suspect this is for grid adaptation to make mesh finer to capture shock waves. In ver.7, grid adaptation related parameters in configuration template file are not used, so it appears that grid adaptation function is dead now.

snow54 commented:
> > For example Green-Gauss gradients have an exclusion for internal boundaries but not for nearfield ones.
> 
> @pcarruscag So, I need to make something similar to internal boundaries? The flowfield looked reasonable though except when I use the VENKATAKRISHNAN limiter where the nearfield boundary seemed to act like half permeable wall.

I am doing some trials after `NEARFIELD_BOUNDARY` is excluded in `computeGradientsGreenGauss` like `INTERNAL_BOUNDARY`. The result looks more reasonable and also the issue I encountered with VENKATAKRISHNAN limiter appears to have been fixed. Can I add changes in this PR or should I create a new PR to address it?

pcarruscag commented:
There are near field boundary conditions in the code. But they are not called. We can cleanup this dead code.
You would need a new PR. If you open one, make sure your branch is in su2code and not in your fork. This will make it easier for us to help (you should have received an invite for the SU2 organizations right?)

snow54 commented:
OK, I will create a new PR. Yes, I have received the invite.
